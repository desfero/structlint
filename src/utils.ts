import partition from "lodash/fp/partition";
import { sep } from "path";
import { debug as debugP } from "debug";
import globby from "globby";
import compact from "lodash/fp/compact";
import curry from "lodash/fp/curry";

import { DeepReadonly, Folder, Demand } from "./types";
import { StructlintError } from "./errors";

const structLintDebug = debugP("struct-lint");

/**
 *  Cast a `value` to exclude `null` and `undefined`.
 *  Throws if either `null` or `undefined` was passed
 */
function nonNullable<T>(
  value: T,
  message?: string,
): asserts value is NonNullable<T> {
  if (value === undefined || value === null) {
    throw new StructlintError(
      message || `Non nullable values expected, received ${value}`,
    );
  }
}

type TOptions = { expandDirectories: boolean };

const getPathFilesAndDirectories = async (
  path: string,
  { expandDirectories }: TOptions,
) => {
  const filesAndFolders = await globby(path, {
    expandDirectories,
    onlyFiles: false,
    gitignore: true,
    markDirectories: true,
  });

  const [directories, files] = partition(
    path => path.endsWith(sep),
    filesAndFolders,
  );

  return { directories, files };
};

/**
 * Check if a given demand is a folder
 */
const isFolder = (
  demand: DeepReadonly<Demand>,
): demand is DeepReadonly<Folder> => "folders" in demand;

/**
 * Get all files of a given demand.
 * If demand is a file itself wraps the file into array
 */
const getFiles = (demand: DeepReadonly<Demand>) => {
  if (!isFolder(demand)) {
    return [demand];
  }

  return compact(Object.values(demand.files));
};

enum IMPORT_TYPE {
  NAMED = "named",
  RELATIVE = "relative",
  ABSOLUTE = "absolute",
}

const getImportType = (importPath: string) => {
  if (importPath.startsWith(".")) {
    return IMPORT_TYPE.RELATIVE;
  }

  if (importPath.startsWith("/")) {
    return IMPORT_TYPE.ABSOLUTE;
  }

  return IMPORT_TYPE.NAMED;
};

const debug = curry((module: string, message: string) =>
  structLintDebug.extend(module)(message),
);

export {
  nonNullable,
  getPathFilesAndDirectories,
  getImportType,
  IMPORT_TYPE,
  debug,
  isFolder,
  getFiles,
};
