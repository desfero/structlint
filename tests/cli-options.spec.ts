import { resolve } from "path";

import { structLint } from "./utils";

describe("cli options", () => {
  it("--version", async () => {
    const { stdout, code } = await structLint({
      cwd: resolve(__dirname, "../examples/violations-js"),
      options: ["--version"],
    });

    expect(code).toBe(0);
    expect(stdout).toMatchSnapshot();
  });

  it("--help", async () => {
    const { stdout, code } = await structLint({
      cwd: resolve(__dirname, "../examples/violations-js"),
      options: ["--help"],
    });

    expect(code).toBe(0);
    expect(stdout).toMatchSnapshot();
  });

  it("--print-config", async () => {
    const { stdout, code } = await structLint({
      cwd: resolve(__dirname, "../examples/violations-js"),
      options: ["--print-config"],
    });

    expect(code).toBe(1);
    expect(stdout).toMatchSnapshot();
  });
});
