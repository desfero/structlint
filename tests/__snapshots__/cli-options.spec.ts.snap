// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli options --help 1`] = `
"Usage: structlint [options]

Project structure linter

Options:
  -V, --version   output the version number
  --print-config  Group violations by configs (default: false)
  -h, --help      display help for command
"
`;

exports[`cli options --print-config 1`] = `
"Resolving configs...
Linting folder structure...
Found 8 violations
-- Config path:  ---------------------------------------------------------------

DISALLOWED IMPORTS: The following imports are disallowed in /modules/transactions/sagas.js
- /components/TxModal
Business logic should never import UI components
- /components/TxSuccessModal
Business logic should never import UI components

DISALLOWED IMPORTS: The following imports are disallowed in /libs/HttpService.js
- lodash
- /modules/utils
- /config/internal

DISALLOWED IMPORTS: The following imports are disallowed in /e2e/investor-flow/invest.spec.e2e.js
- /libs/HttpService
- /e2e/investor-flow/payout.spec.e2e

-- Config path: components -----------------------------------------------------

DISALLOWED IMPORTS: The following imports are disallowed in /components/shared/Button.jsx
- /components/TxSuccessModal
"
`;

exports[`cli options --version 1`] = `
"1.0.0-beta.0
"
`;
